### PR TITLE
Allowing nested username (e.g. /rhat/smahajan/) in docker tag command

### PR DESCRIPTION
**Pull request by shishir-a412ed**
Closed #11115 because of a bad rebase.
Opened a new one here.

### CODE REVIEW

**Comment by jfrazelle** [comment]
please next time force push your branch instead of opening a new PR

**Comment by shishir-a412ed** [comment]
@jfrazelle noted :)

**Comment by jfrazelle** [comment]
ping @dmcgowan @dmp42 this will work right also @kencochrane for hub

Also @tiborvass do you think this will effect anything with namespacing

**Comment by tiborvass** [comment]
@jfrazelle I'm afraid this one will take some time to review, and it is not only because of code review.
I'm putting this in design review.

@shishir-a412ed 

Some questions are:
- if the hub does not support it currently, will it? 
- is there a usecase for this without the hub supporting it?
- If so, would it be okay to accept naming images like that, and erroring out on push to a docker registry until the registry supports it?

File: registry/config.go [hunk 1]
```diff
@@ -32,7 +32,7 @@ const (
 var (
 	ErrInvalidRepositoryName = errors.New("Invalid repository name (ex: \"registry.domain.tld/myrepos\")")
 	emptyServiceConfig       = NewServiceConfig(nil)
-	validNamespaceChars      = regexp.MustCompile(`^([a-z0-9-_]*)$`)
+	validNamespaceChars      = regexp.MustCompile(`^([a-z0-9-_/]*)$`)
```
Comment (ncdc) [inline registry/config.go]: FYI, this is what the v2 registry currently does:

- component regex: https://github.com/docker/distribution/blob/master/registry/api/v2/names.go#L38
- full name regex: https://github.com/docker/distribution/blob/master/registry/api/v2/names.go#L46

**Comment by ncdc** [comment]
@shishir-a412ed can you add some test cases?

File: registry/config.go [hunk 2]
```diff
@@ -207,18 +207,20 @@ func validateRemoteName(remoteName string) error {
...
 			return fmt.Errorf("Invalid repository name (%s), cannot specify 64-byte hexadecimal strings", name)
 		}
 	} else {
-		namespace = nameParts[0]
-		name = nameParts[1]
+		namespace = remoteName[:lastIndex]
+		name = remoteName[lastIndex+1:]
 	}
 	if !validNamespaceChars.MatchString(namespace) {
 		return fmt.Errorf("Invalid namespace name (%s). Only [a-z0-9-_] are allowed.", namespace)
```
Comment (ncdc) [inline registry/config.go]: Need to update the error message text

Comment (shishir-a412ed) [inline registry/config.go]: @ncdc Will add some test cases and fix the error message.

**Comment by dmcgowan** [comment]
The v1 registry has no support for nested usernames.  Currently this is updating an already messy implementation verifying the namespace.  This code will be refactored for 1.7 when the code is split out into distribution. I am marking it as 1.7.0, we might close it with an alternative implementation later which only works for v2 registries and enforces v1 namespace rules.

**Comment by tiborvass** [comment]
@dmcgowan apart from not wanting to change messy code that is going to be refactored, can you comment on the design itself? Do you agree? + https://github.com/docker/docker/pull/11116#issuecomment-77172790

@ncdc thanks for helping with the review! Please remember that this is in design review phase :) It's in order to avoid unnecessary back-and-forths between contributors and maintainers.

**Comment by stevvooe** [comment]
This should be handled through the validation of names via the docker/distribution imports. While the library support is fresh, the validation and regular expressions are solid and should be used here.

Note that we are going away from a limited `<namespace>/<name>` view, so I am unsure what problem this specific PR is addressing. To clarify, the weak notion of naming the path components should be dropped.

It would be helpful if you could outline the desired behavior clearly in the the description, including examples of what was valid/invalid before and after this change. From the looks of it, we may be already supporting what you're looking for.

**Comment by shishir-a412ed** [comment]
@stevvooe 
Example: registry.access.redhat.com:5000/rhat/dockeruser/smahajan/busybox:latest
In this example: namespace (username) is rhat/dockeruser/smahajan

Before this change:
This is an invalid namespace (username) if I try to run the `docker tag` command.

We want to allow this to have fully qualified usernames. This change fixes that.
If this is already being supported, let me know and I ll test it.

**Comment by stevvooe** [comment]
@shishir-a412ed I would think this should already be supported, if not already. I think the confusion I have with this PR is that it makes the slash (`/`) a valid namespace component character, which it absolutely is not.

Check out https://github.com/docker/distribution/blob/master/registry/api/v2/names.go for "the future of naming". We are moving towards that layout as much as possible and the tags command should probably reflect that.

Let me know what your findings are.

**Comment by dmcgowan** [comment]
This implementation should match the rules defined in the distribution package, look at the link's Andy sent.

As to the comment from @stevvooe, it would be better if the name was split apart and each component individually checked.  The way it is coded now names such as "docker/-bad/docker" will be considered valid.

**Comment by shishir-a412ed** [comment]
This change uses #11478 . Waiting for it to be merged.

**Comment by tiborvass** [comment]
@shishir-a412ed please close this PR then, and reopen it once the dependent PR is merged.

**Comment by icecrime** [comment]
Yes, I'm sorry but there is just to much to keep track of to afford dealing with PR that need to wait on other PR.

**Comment by rhatdan** [comment]
Since https://github.com/docker/docker/pull/11478 has been merged, should we reopen this pull request.

**Comment by ncdc** [comment]
@rhatdan this is superseded by #13070

**Comment by rhatdan** [comment]
Thanks, tough to follow the bouncing pull requests.

**Comment by shishir-a412ed** [comment]
@rhatdan apparently github doesn't allow to reopen the PR, if we force push on a closed PR.
and a force push is sometimes needed to rebase. So had to open a new PR.

https://github.com/isaacs/github/issues/361

**Pull request closed without merging**